### PR TITLE
feat(parser): extract authors from CFF files

### PR DESCRIPTION
**Pull request by rmfranken**
This branch makes it possible to retrieve "author" objects from a CFF file and write them to turtle. It slightly refactored the generic parse function to output a graph structure (triples) instead of properties and values ("doubles?").

Additionally, minor fixes in poetry dependencies.

### CODE REVIEW

**Comment by rmfranken** [comment]
@cmdoret I'm especially interested in the tests - how do you generally decide how "deep" to test? Looking at the other tests I saw the level detail was not super high, so I tried to match that in my test- but happy to hear if there is a better way of deciding that.

Also, I think the container registry action that is failing seems to be related to a version number? Is that something I should add to my PR somehow? It seems related to the version of git, rather than any code I added :/

File: gimie/parsers/__init__.py [hunk 1]
```diff
@@ -24,6 +24,8 @@
 from gimie.parsers.license import LicenseParser, is_license_filename
 from gimie.parsers.cff import CffParser
 
+from rdflib import Graph, URIRef
```
Comment (cmdoret) [inline gimie/parsers/__init__.py:27]:
```suggestion
from rdflib import Graph
```

fix: unused import

**Review by cmdoret — CHANGES_REQUESTED** [review]
Nice work! I have a bunch of suggestions to improve code readability.

About tests: I would just add a test for behaviour on a cff with broken orcid (not a uri) and authors without doi. And making sure they don't appear in the output.

> Note: you don't need to create cff files for tests, since the functions take bytes, you can write plain yaml in the test, e.g.:
```py
data = b"""
<yaml-content>
"""
get_cff_authors(data)
```

I'm not sure there's an absolute rule about test depths, but in general, it's good to cover the obvious edge cases, and have them run on important code paths.

File: gimie/parsers/cff.py [hunk 2]
```diff
@@ -18,31 +18,68 @@
...
         if doi:
-            props.add((SDO.citation, URIRef(doi)))
-        return props
+            rdf_graph.add((self.subject, SDO.citation, URIRef(doi)))
+        if not authors:
+            return rdf_graph
+        for author in authors:
+            if author["orcid"]:
+                rdf_graph.add(
+                    (self.subject, SDO.author, URIRef(author["orcid"]))
```
[the remark this answers is not in the document]
Comment (rmfranken) [reply gimie/parsers/cff.py:56]: assigning once I like, but the URIRef I think still makes sense to do at the .add step, as I will need it as a string in one case, and as a URIRef in the other.

File: tests/test_output.py [hunk 3]
```diff
@@ -35,22 +35,22 @@ def test_validate_output_is_linked_data():
...
-        abort_on_first=False,
-        allow_infos=False,
-        allow_warnings=False,
-        meta_shacl=False,
-        advanced=False,
-        js=False,
-        debug=False,
-    )
-    assert valid_graph
+# @pytest.mark.skip("not yet implemented")
```
[the remark this answers is not in the document]
Comment (rmfranken) [reply tests/test_output.py]: fixed by yeeting pyshacl out completely

**Comment by rmfranken** [comment]
Since I was making a orcid matcher, I thought I would also move the doi matching logic to utils. Does that make sense?

File: tests/test_cff.py [hunk 4]
```diff
@@ -8,3 +10,76 @@ def test_parse_cff():
...
+        .parse(data=cff_file_broken_orcid)
+        .serialize(format="ttl")
+        == "\n"
+    )
+    assert (
+        CffParser(subject=URIRef("https://example.org/"))
+        .parse(data=cff_file_author_without_orcid)
+        .serialize(format="ttl")
+        == "\n"
+    )
```
Comment (cmdoret) [inline tests/test_cff.py:74]:
```suggestion
    assert len(
        CffParser(subject=URIRef("https://example.org/"))
        .parse(data=cff_file_emptyish)
        ) == 0
    
    assert len(
        CffParser(subject=URIRef("https://example.org/"))
        .parse(data=cff_file_bad_syntax)
    ) == 0
    
    assert len(
        CffParser(subject=URIRef("https://example.org/"))
        .parse(data=cff_file_broken_orcid)
    ) == 0
    
    assert len(
        CffParser(subject=URIRef("https://example.org/"))
        .parse(data=cff_file_author_without_orcid)
    ) == 0
```
**comment:** relying on serialization conventions is kinda brittle (they can change the newline thing).
**suggestion:** actually test expectation on the graph structure itself

Comment (cmdoret) [reply tests/test_cff.py:74]: **comment(non-blocking):** as the test logic is constant, if you want to make this even neater, you could use pytest fixtures instead to make a parametrized test. See e.g. https://github.com/sdsc-ordes/gimie/blob/1866420f39b15312e6f73cab60c73f8e3ca677d3/tests/test_github.py#L16

File: tests/test_cff.py [hunk 5]
```diff
@@ -8,3 +10,76 @@ def test_parse_cff():
...
+
+
+def test_doi():
```
Comment (cmdoret) [inline tests/test_cff.py:77]:
```suggestion
def test_parse_doi():
```
**suggestion:** name after action

File: tests/test_cff.py [hunk 6]
```diff
@@ -8,3 +10,76 @@ def test_parse_cff():
...
+    cff_file = b"""
+    cff-version: 1.2.0
+    title: gimie
+    doi: 10.5281/zenodo.1234567
+    """
+    assert "https://doi.org/10.5281/zenodo.1234567" in CffParser(
+        subject=URIRef("https://example.org/")
+    ).parse(data=cff_file).serialize(format="ttl")
```
Comment (cmdoret) [inline tests/test_cff.py:85]: **suggestion(non-blocking):** testing if the DOI is among objects would be more robust than in the serialization, also it lets you check if it's indeed a Literal and not URIRef.

Comment (cmdoret) [reply tests/test_cff.py:85]: prob something like

```py
assert Literal("https://...7") in CffParser(...).parse(...).objects()
```

**Review by cmdoret — COMMENTED** [review]
Nice, this is starting to look really good! I have a few suggestions to improve things but most of them are optional.

The only really important one is the orcid url check in CffParser.

**Comment by cmdoret** [comment]
**note:** bumped the dockerfile base layer so that ci succeeds

File: gimie/parsers/cff.py [hunk 7]
```diff
@@ -40,22 +41,25 @@ def parse(self, data: bytes) -> Graph:
...
         if not authors:
-            return rdf_graph
+            return extracted_cff_triples
         for author in authors:
-            if author["orcid"]:
-                rdf_graph.add(
-                    (self.subject, SDO.author, URIRef(author["orcid"]))
+            orcid = URIRef(author["orcid"])
+            if is_valid_orcid(orcid):
+                extracted_cff_triples.add(
```
[the remark this answers is not in the document]
Comment (rmfranken) [reply gimie/parsers/cff.py:57]: That makes total sense - this is much better.

**Review by cmdoret — APPROVED** [review]
LGTM :rocket: well done!
